### ISSUE TITLE
Improve Vector documentation

### ISSUE THREAD

**Issue by nwinter** [codecombat/codecombat#3110]
See [this thread](http://discourse.codecombat.com/t/vectors-and-help/5226). 

In addition to writing Vector docs examples for other languages (see https://github.com/codecombat/codecombat/issues/1237 for more), we also need to figure out what we're doing in terms of which Vector methods should have their examples shown as class methods and which should be instance methods. Probably our code related to `owner` needs to get smarter. Part of it is [here](https://github.com/codecombat/codecombat/blob/master/app/views/play/level/tome/DocFormatter.coffee#L132-L132).

**Comment by Driphter** [comment]
Is there any reason that any of the Vector methods couldn't be both instance methods as well as class methods? This would remove the confusion referred to in the thread as well as allow for different styles depending on the situation.

``` javascript
var diff = Vector.subtract(vecA, vecB); // I prefer this one
var diff = vecA.subtract(vecB);

var vec = Vector.add(Vector.multiply(Vector.normalize(Vector.subtract(vecA, vecB)), distance), vecA);
var vec = vecA.subtract(vecB).normalize().multiply(distance).add(vecA); // I prefer this one.
```

**Comment by nwinter** [comment]
Hmm. Those methods already exist and are mutating, used all throughout the game engine. But maybe in player code we could somehow hook up non-mutating versions, then allowing all the instance methods?

**Comment by Driphter** [comment]
That's what I was thinking. I didn't realize until now that the instance methods mutate the Vector. That could be confusing, especially when the class methods don't mutate. The docs currently specifically state that a new Vector is returned as well.

**Comment by Driphter** [comment]
Would it be reasonable to extend Vector with a class that only contains non-mutating methods? If so, any ideas for a good name for the class?

**Comment by nwinter** [comment]
All the docs and sample code and global Aether imports and stuff refers to Vectors, so I wouldn't want to change anything's name. I would just want to find a quick hack to make player-code use non-mutating methods instead.

**Comment by Driphter** [comment]
Couldn't we just change [this line](https://github.com/codecombat/codecombat/blob/5095eac4ac791165527b62b401886ba1ef446510/app/lib/aether_utils.coffee#L1) from:
```Aether.addGlobal 'Vector', require './world/vector'```
to:
```Aether.addGlobal 'Vector', require './world/playerVector'```
?

**Comment by nwinter** [comment]
I haven't tested but I think you will end up with player vectors getting into game engine code if that's all you do, and game engine code expects things to be mutable. I suppose it would be slightly possible to refactor all game engine code for immutability in one heroic push, but I'm looking for a way that the current Vectors (no matter where they are created) can just know whether they currently are in player code and decide whether to mutate or not based on that, so we wouldn't have to do a lot of work.

**Comment by Driphter** [comment]
I agree, refactoring the engine code not only be a pain, but a performance decrease as well. Maybe we could add the ability to use an object for apiProperties and friends? 

Something like:
{'apiMethodName': 'internalMethodName',
 'anotherMethodName': (x) -> foo(x)}

**Comment by nwinter** [comment]
Wait, a Vector can already see if it is currently a normal Vector or an API-protected clone, right? Maybe the Vector code itself can just detect this, in each method, and skip mutation if it detects the API protection (like, if it has the `__aetherAPIValue` property).

If that works, you have the `if @__aetherAPIValue` check in each method, but that's not a big performance hit, and is probably good enough for now.

I don't have time to test this out soon, and it will require changes to Vector.coffee to try. I forget, did you have the dev setup working okay on your machine?

**Comment by Driphter** [comment]
Yep, everything is set up and working. I can mess around with it today.
